prExplorer/client-app/src/app/stores/productStore.ts

```diff
@@ -0,0 +1,125 @@
+import { makeAutoObservable, runInAction } from "mobx";
+import { Product } from "../models/product";
+import agent from "../api/agent";
+import { v4 as uuid } from "uuid";
+
+export default class ProductStore {
+  productRegistry = new Map<string, Product>();
+  selectedProduct: Product | undefined = undefined;
+  editMode = false;
+  loading = false;
+  loadingInitial = true;
+
+  constructor() {
+    makeAutoObservable(this);
+  }
+
+  get productsSortedByName() {
+    const productList = Array.from(this.productRegistry.values());
+    productList.sort((a, b) => {
+      const titleA = a.title.toLowerCase();
+      const titleB = b.title.toLowerCase();
+      return titleA.localeCompare(titleB);
+    });
+    return productList;
+  }
+
+  loadProducts = async () => {
+    this.setLoadingInitial(true);
+    try {
+      const products = await agent.Products.list();
+      products.forEach((product) => {
+        this.setProduct(product);
+      });
+      this.setLoadingInitial(false);
+    } catch (error) {
+      console.log(error);
+      this.setLoadingInitial(false);
+    }
+  };
+
+  loadProduct = async (id: string) => {
+    let product = this.getProduct(id);
+    if (product) {
+      this.selectedProduct = product;
+      return product;
+    } else {
+      this.loadingInitial = true;
+      try {
+        product = await agent.Products.details(id);
+        this.setProduct(product);
+        runInAction(() => this.selectedProduct = product);
+        this.setLoadingInitial(false);
+        return product;
+      } catch (error) {
+        console.log(error);
+        this.setLoadingInitial(false);
+      }
+    }
+  };
+
+  private setProduct = (product: Product) => {
+    this.productRegistry.set(product.id, product);
+  };
+
+  private getProduct = (id: string) => {
+    return this.productRegistry.get(id);
+  };
+
+  setLoadingInitial = (state: boolean) => {
+    this.loadingInitial = state;
+  };
+
+  createProduct = async (product: Product) => {
+    this.loading = true;
+    product.id = uuid();
+    try {
+      await agent.Products.create(product);
+      runInAction(() => {
+        this.productRegistry.set(product.id, product);
+        this.selectedProduct = product;
+        this.editMode = false;
+        this.loading = false;
+      });
+    } catch (error) {
+      console.log(error);
+      runInAction(() => {
+        this.loading = false;
+      });
+    }
+  };
+
+  updateProduct = async (product: Product) => {
+    this.loading = true;
+    try {
+      await agent.Products.update(product);
+      runInAction(() => {
+        this.productRegistry.set(product.id, product);
+        this.selectedProduct = product;
+        this.editMode = false;
+        this.loading = false;
+      });
+    } catch (error) {
+      console.log(error);
+      runInAction(() => {
+        this.loading = false;
+      });
+    }
+  };
+
+  deleteProduct = async (id: string) => {
+    this.loading = true;
```

Comment:
I've already suggested that in another PR - it would be better if `loading` state is handled with separate actions, not inside `deleteProduct` or `updateProduct`. Then the consumer of this `productStore` can decide whether it wants to start/stop loading. This is not a big deal, but could be useful to have such a habit ;)